### ISSUE TITLE
support for kotlin

### ISSUE THREAD

**Issue by abrams27** [build-server-protocol/build-server-protocol#520]
### Describe the use-case for this feature

we'd like to support kotlin in our plugin, and in order to do that we need to obtain some info about the used sdk in the project

### What do you propose

in the near future we will create a kotlin extension and PR with changes, probably we will need only additional data field with sdk info, and maybe some info about relationship between targets (are they in the same kotlin module - e.g. bazel `associates` https://bazelbuild.github.io/rules_kotlin/kotlin.html#kt_jvm_library) (but we need to investigate it a bit anyway)

### I've checked to make sure there isn't already a way to support this in the current protocol

- [X] I've checked the existing protocol and there isn't a way to do this currently

### Maintainer approval (This is for the maintainers)

- [ ] Jetbrains
- [ ] Metals
- [ ] Scala Center

**Comment by Arthurm1** [comment]
Is the Kotlin extension spec [here](https://github.com/JetBrains/bazel-bsp/blob/ec256a99f85e1df882439308a6dbeb6951e28cd2/protocol/src/main/kotlin/org/jetbrains/bsp/KotlinBuildTarget.kt)?  And is that pretty much final?

Also - what exactly is in the `associates` field and how does it differ from `BuildTarget#dependencies`.  The Go embed link seems to suggest it's packaging related - so is this for some kind of fat jar setup?  And how would a BSP client use this information - I would have thought that would all be hidden in the build tool/server and the client wouldn't care.

**Comment by agluszak** [comment]
@abrams27

**Comment by abrams27** [comment]
`associates` means that the targets will be compiled together in one "module" so you can use symbols with `internal` visibility (https://bazelbuild.github.io/rules_kotlin/kotlin#kt_jvm_library) on the client side this info is required (at least in intellij) since we map 1-1 build targets so it's required to handle `internal` visibility properly

**Comment by lefou** [comment]
This is interesting. How do you provide `associates` to the Kotlin compiler? I maintain a Kotlin plugin for Mill, and I'm not aware of such option.

**Comment by lefou** [comment]
Yeah, I guess I'm actually more interested in the server side then. It looks a bit strange to me to control the friendship between modules via a keyword from the one side, but via a compiler setting from the other.

**Comment by abrams27** [comment]
yes but there are cases when it's a "fictional" thing - for example library and test, in bazel it's impossible to compile them together, but the test in the same package should be able to see the symbols, and the "module" split caused by the design of bazel should not block it (it's similar in gradle support in intellij)

**Comment by lefou** [comment]
@abrams27 Yeah, that makes sense. Out of curiosity: Do you need to explicitly mark those `associate` modules in Bazel or it this some implicit setup for tests? Can you share an example? If this is some automatic thing, it might make sense to also implement it in mill-kotlin.

**Comment by abrams27** [comment]
you need to set it explicitly,
we have a bit of them in our tests: https://github.com/JetBrains/bazel-bsp/blob/ec256a99f85e1df882439308a6dbeb6951e28cd2/executioncontext/workspacecontext/src/test/kotlin/org/jetbrains/bsp/bazel/workspacecontext/BUILD#L7
